From 9f49f06e91e80a7823d34702e36c1f61c50f8db2 Mon Sep 17 00:00:00 2001
From: Christophe Roullier <christophe.roullier@foss.st.com>
Date: Tue, 21 May 2024 17:32:30 +0200
Subject: [PATCH 1134/1141] net: phy: smsc: Fix wol config NULL pointer
 dereferencing

In case of a suspend resume flow config_init() function is called.
If the network interface have not been set up before the suspend resume
the attached_dev pointer is NULL.
config_init() call the smsc_phy_config_wol() function which dereference
attached_dev. This cause this oops kernel NULL pointer dereference:
[   33.682809] Unable to handle kernel NULL pointer dereference at virtual address 000001d4
[   33.690893] [000001d4] *pgd=00000000
[   33.694490] Internal error: Oops: 5 [#1] PREEMPT SMP ARM
[   33.699754] Modules linked in: ...
[   33.740777] CPU: 0 PID: 186 Comm: sh Not tainted 6.1.28 #1
[   33.746141] Hardware name: STM32 (Device Tree Support)
[   33.751295] PC is at smsc_phy_config_wol+0x14/0xac
[   33.756065] LR is at smsc_phy_config_init+0x70/0x74

This patch fixes the issue by returning -ENODEV if attached_dev is NULL.

Signed-off-by: Kory Maincent <kory.maincent@bootlin.com>
Signed-off-by: Christophe Roullier <christophe.roullier@foss.st.com>
Change-Id: I520baee3384398946e79a78d1c4858402b9b5879
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/381420
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gwenael TREUVEUR <gwenael.treuveur@st.com>
---
 drivers/net/phy/smsc.c | 3 +++
 1 file changed, 3 insertions(+)

--- a/drivers/net/phy/smsc.c
+++ b/drivers/net/phy/smsc.c
@@ -108,6 +108,9 @@ static int smsc_phy_config_wol(struct ph
 	int i, wol_ctrl, wol_filter;
 	u16 pwd[3] = {0, 0, 0};
 
+	if (!phydev->attached_dev)
+		return -ENODEV;
+
 	/* Write @MAC in LAN8742_MMD3_MAC_ADDRA/B/C registers */
 	const u8 *mac_addr = phydev->attached_dev->dev_addr;
 	/* Store the device address for the magic packet */
